### PR TITLE
refactor(cna): make `create-next-app` even smaller and faster

### PR DESCRIPTION
**Pull request by SukkaW**
The PR further reduces the `create-next-app` installation size by another 80 KiB:

- Replace the callback version of Node.js built-in `dns` API usage with `dns/promise` + async/await
- Replace `got` w/ `fetch` since Next.js and `create-next-app` now target Node.js 18.17.0+
- Download and extract the tar.gz file in the memory (without creating temporary files). This improves the performance.
- Some other minor refinements.

Following these changes, the size of `dist/index.js` is now 536 KiB.

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js refactor-cna-1 | Change |
| - | - | - | - |
| buildDuration | 10.7s | 10.7s | N/A |
| buildDurationCached | 6.2s | 6.1s | N/A |
| nodeModulesSize | 175 MB | 175 MB | ✓ |
| nextStartRea..uration (ms) | 420ms | 401ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js refactor-cna-1 | Change |
| - | - | - | - |
| 199-HASH.js gzip | 30.4 kB | 30.5 kB | N/A |
| 3f784ff6-HASH.js gzip | 53.2 kB | 53.2 kB | ✓ |
| 494.HASH.js gzip | 182 B | 182 B | ✓ |
| framework-HASH.js gzip | 45.5 kB | 45.5 kB | ✓ |
| main-app-HASH.js gzip | 254 B | 252 B | N/A |
| main-HASH.js gzip | 33 kB | 33 kB | N/A |
| webpack-HASH.js gzip | 1.75 kB | 1.75 kB | N/A |
| Overall change | 98.9 kB | 98.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js refactor-cna-1 | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js refactor-cna-1 | Change |
| - | - | - | - |
| _app-HASH.js gzip | 205 B | 205 B | ✓ |
| _error-HASH.js gzip | 182 B | 181 B | N/A |
| amp-HASH.js gzip | 505 B | 507 B | N/A |
| css-HASH.js gzip | 322 B | 323 B | N/A |
| dynamic-HASH.js gzip | 2.59 kB | 2.59 kB | N/A |
| edge-ssr-HASH.js gzip | 258 B | 259 B | N/A |
| head-HASH.js gzip | 348 B | 347 B | N/A |
| hooks-HASH.js gzip | 369 B | 368 B | N/A |
| image-HASH.js gzip | 4.38 kB | 4.38 kB | N/A |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 2.71 kB | 2.7 kB | N/A |
| routerDirect..HASH.js gzip | 318 B | 318 B | ✓ |
| script-HASH.js gzip | 384 B | 383 B | N/A |
| withRouter-HASH.js gzip | 319 B | 320 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 885 B | 885 B | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js refactor-cna-1 | Change |
| - | - | - | - |
| _buildManifest.js gzip | 483 B | 484 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js refactor-cna-1 | Change |
| - | - | - | - |
| index.html gzip | 526 B | 528 B | N/A |
| link.html gzip | 540 B | 540 B | ✓ |
| withRouter.html gzip | 522 B | 523 B | N/A |
| Overall change | 540 B | 540 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js refactor-cna-1 | Change |
| - | - | - | - |
| edge-ssr.js gzip | 96.7 kB | 96.7 kB | N/A |
| page.js gzip | 141 kB | 141 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js refactor-cna-1 | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 625 B | 627 B | N/A |
| middleware-r..fest.js gzip | 148 B | 151 B | N/A |
| middleware.js gzip | 24.8 kB | 24.8 kB | N/A |
| edge-runtime..pack.js gzip | 1.92 kB | 1.92 kB | ✓ |
| Overall change | 1.92 kB | 1.92 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js refactor-cna-1 | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 167 kB | 167 kB | ✓ |
| app-page-exp..prod.js gzip | 94.5 kB | 94.5 kB | ✓ |
| app-page-tur..prod.js gzip | 95.2 kB | 95.2 kB | ✓ |
| app-page-tur..prod.js gzip | 89.9 kB | 89.9 kB | ✓ |
| app-page.run...dev.js gzip | 137 kB | 137 kB | ✓ |
| app-page.run..prod.js gzip | 89.2 kB | 89.2 kB | ✓ |
| app-route-ex...dev.js gzip | 23.5 kB | 23.5 kB | ✓ |
| app-route-ex..prod.js gzip | 16.4 kB | 16.4 kB | ✓ |
| app-route-tu..prod.js gzip | 16.4 kB | 16.4 kB | ✓ |
| app-route-tu..prod.js gzip | 16 kB | 16 kB | ✓ |
| app-route.ru...dev.js gzip | 22.9 kB | 22.9 kB | ✓ |
| app-route.ru..prod.js gzip | 16 kB | 16 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.37 kB | 9.37 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.64 kB | 9.64 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.37 kB | 9.37 kB | ✓ |
| pages-turbo...prod.js gzip | 21.5 kB | 21.5 kB | ✓ |
| pages.runtim...dev.js gzip | 22.2 kB | 22.2 kB | ✓ |
| pages.runtim..prod.js gzip | 21.5 kB | 21.5 kB | ✓ |
| server.runti..prod.js gzip | 48.5 kB | 48.5 kB | ✓ |
| Overall change | 926 kB | 926 kB | ✓ |

</details>

</details>
Commit: eb977dd53ae1c22e449139f1a200b0b1dc544c29

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 6426bd577d8feda23aa22dbfd92ecca55a4a7acd

`pnpm test-dev test/e2e/app-dir/navigation/navigation.test.ts` 
- app dir - navigation > query string > useParams identity between renders > should be stable in pages

<details>
<summary>Expand output</summary>

  ● app dir - navigation › query string › useParams identity between renders › should be stable in pages

    TIMED OUT: success

    fail

    undefined

      626 |
      627 |   if (hardError) {
    > 628 |     throw new Error('TIMED OUT: ' + regex + '\n\n' + content + '\n\n' + lastErr)
          |           ^
      629 |   }
      630 |   return false
      631 | }

      at check (lib/next-test-utils.ts:628:11)
      at runTests (e2e/app-dir/navigation/navigation.test.ts:77:11)
      at Object.<anonymous> (e2e/app-dir/navigation/navigation.test.ts:102:11)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

File: packages/create-next-app/create-app.ts [hunk 1]
```diff
@@ -132,7 +131,7 @@ export async function createApp({
 
   const appName = path.basename(root)
 
-  await makeDir(root)
+  fs.mkdirSync(root, { recursive: true })
```
Comment (styfle) [inline packages/create-next-app/create-app.ts:134]: Any reason this was changed from async to sync?

File: packages/create-next-app/helpers/examples.ts [hunk 2]
```diff
@@ -17,8 +11,15 @@ export type RepoInfo = {
...
 
 export async function isUrlOk(url: string): Promise<boolean> {
-  const res = await got.head(url).catch((e) => e)
-  return res.statusCode === 200
+  try {
+    const res = await fetch(url, { method: 'HEAD' })
+    if (res.status !== 200) {
+      return false
+    }
+    return true
```
Comment (styfle) [inline packages/create-next-app/helpers/examples.ts:19]:
```suggestion
    return res.status === 200
```

File: packages/create-next-app/helpers/examples.ts [hunk 3]
```diff
@@ -82,50 +88,50 @@ export function existsInRepo(nameOrUrl: string): Promise<boolean> {
...
-  await pipeline(got.stream(url), createWriteStream(tempFile))
-  return tempFile
+async function downloadTarStream(url: string) {
+  const resp = await fetch(url)
+
+  if (!resp.body) {
+    throw new Error(`Failed to download: ${url}`)
+  }
+
+  return Readable.fromWeb(resp.body as import('stream/web').ReadableStream)
```
Comment (styfle) [inline packages/create-next-app/helpers/examples.ts:98]:
```suggestion
  const res = await fetch(url)

  if (!res.body) {
    throw new Error(`Failed to download: ${url}`)
  }

  return Readable.fromWeb(res.body as import('stream/web').ReadableStream)
```

File: packages/create-next-app/templates/index.ts [hunk 4]
```diff
@@ -118,7 +117,7 @@ export const installTemplate = async ({
   }
 
   if (srcDir) {
-    await makeDir(path.join(root, 'src'))
+    await fs.mkdir(path.join(root, 'src'))
```
Comment (styfle) [inline packages/create-next-app/templates/index.ts:120]: Do we need to add `recursive: true` here too?

File: packages/create-next-app/create-app.ts (hunk 1 above)
Comment (SukkaW) [reply packages/create-next-app/create-app.ts:135]: @styfle `create-next-app` is just a simple single-thread script-like app that doesn't have multiple concurrent promises.

I've explained that before: https://github.com/vercel/next.js/pull/56917#discussion_r1362421015

File: packages/create-next-app/templates/index.ts (hunk 4 above)
Comment (SukkaW) [reply packages/create-next-app/templates/index.ts]: Not sure `recursive: true` is necessary here, given that the root should already exist and only a one-level src folder will be created.

Nonetheless, including it won't cause any harm. Addressed in [`786162b` (#58030)](https://github.com/vercel/next.js/pull/58030/commits/786162b26394a37fcf1a2d63c3cedc99a5efb911).

File: packages/create-next-app/helpers/examples.ts (hunk 3 above)
Comment (SukkaW) [reply packages/create-next-app/helpers/examples.ts:95]: Addressed in [`786162b` (#58030)](https://github.com/vercel/next.js/pull/58030/commits/786162b26394a37fcf1a2d63c3cedc99a5efb911).

File: packages/create-next-app/helpers/examples.ts (hunk 2 above)
Comment (SukkaW) [reply packages/create-next-app/helpers/examples.ts]: Addressed in [`786162b` (#58030)](https://github.com/vercel/next.js/pull/58030/commits/786162b26394a37fcf1a2d63c3cedc99a5efb911).

**Review by styfle — APPROVED** [review]
Great, thanks!

**Comment by SukkaW** [comment]
@styfle I just rebased the PR and fixed the conflicts inside the `pnpm-lock.yaml`.

**Review by styfle — APPROVED** [review]
Great work, thanks! 🎉

**Pull request merged**